core/lib/merkle_tree/src/pruning.rs

```diff
@@ -94,17 +111,24 @@ impl<DB: PruneDatabase> MerkleTreePruner<DB> {
         self.poll_interval = poll_interval;
     }
 
-    fn target_retained_version(&self) -> Option<u64> {
+    /// Returns max version number that can be safely pruned, so that after pruning there is at least one version present after pruning
+    pub fn last_prunable_version(&self) -> Option<u64> {
         let manifest = self.db.manifest()?;
-        let latest_version = manifest.version_count.checked_sub(1)?;
-        latest_version.checked_sub(self.past_versions_to_keep)
+        manifest.version_count.checked_sub(1)
     }
 
     #[doc(hidden)] // Used in integration tests; logically private
     #[allow(clippy::range_plus_one)] // exclusive range is required by `PrunePatchSet` constructor
-    pub fn run_once(&mut self) -> Option<PruningStats> {
-        let target_retained_version = self.target_retained_version()?;
+    pub fn prune_up_to(&mut self, target_retained_version: u64) -> Option<PruningStats> {
         let min_stale_key_version = self.db.min_stale_key_version()?;
+
+        //We must leave at least one version
+        let last_prunable_version = self.last_prunable_version();
+        if last_prunable_version.is_none() {
+            tracing::info!("Nothing to prune; skipping");
+            return None;
+        }
+        let target_retained_version = min(target_retained_version, last_prunable_version.unwrap());
```

Comment:
Nit: You can use let-else or `?` on `last_prunable_version` for better code structuring and to avoid the `unwrap()` later.

Reply:
I deliberately wanted to avoid let-else as I generally try to keep indentation level as low as possible for bigger blocks of code, ? is a good suggestion